File: src/core/expr/code/env.rb
#=begin
#This file stores various types of environments

#@parent is the parent env used when the local lookup fails
#@parent is never written to, even when a variable is already defined there
#=end

require 'enso'

module Env
  module BaseEnv
    def set!(key, &block)
      self[key] = block.call(self[key])
    end
    
    def set(key, &block)
      res = self.clone
      res.set!(key, block)
      res
    end
    
    def set_parent(env)
      @parent = env
      self
    end
    
    def set_grandparent(env)
      if @parent.nil? || @parent=={}
        set_parent(env)
      else
        @parent.set_grandparent(env)
      end
    end
    
    def to_s
      r = []
      each {|k,v| r << "#{k}=>#{v}"}
      "{ #{r.join(", ")} }"
    end
  end

  class HashEnv
    include BaseEnv
    def initialize(hash={}, parent=nil)
      super()
      @hash=hash
      @parent=parent
    end
    
    def [](key)
      key = key.to_s
      if @hash.has_key?(key)
        @hash[key]
      else
        @parent && @parent[key]
      end
    end

    def []=(key, value)
      key = key.to_s
      if @hash.has_key?(key) #if defined in current env
        @hash[key] = value
      elsif @parent && @parent.has_key?(key) #if defined in parent env
        @parent[key] = value
      else #new variable goes into current env
        @hash[key] = value
      end
    end
    
    def has_key?(key)
      key = key.to_s
      @hash.has_key?(key) || (@parent && @parent.has_key?(key))
    end
    
    def keys
      (@hash.keys + (@parent.nil? ? [] : @parent.keys)).uniq
    end
      
    def to_s
      "#{@hash.to_s}-#{@parent}"
    end
  end

  #Env that simulates an MObject
  class ObjEnv
    include BaseEnv
  
    attr_reader :obj

    def initialize(obj, parent = nil)
      super()
      @obj = obj
      @parent = parent
    end
    
    def [](key)
      if key == "self"
        #puts "SELF = #{obj}"
        @obj
      elsif @obj.schema_class.all_fields.any?{|f|f.name == key}
        @obj[key]
      else
        @parent && @parent[key]
      end
    end
    
    def []=(key, value)
      begin
        @obj[key] = value
      rescue
        @parent && @parent[key] = value
      end
    end
    
    def has_key?(key)
      key == "self" || @obj.schema_class.all_fields[key] || (@parent && @parent.has_key?(key))
    end
        
    def keys
      (@obj.schema_class.all_fields.keys + (@parent.nil? ? [] : @parent.keys)).uniq
    end
      
    def to_s
      "#{@obj.to_s}-#{@parent}"
    end
    
    def type(fname)
      x = @obj.schema_class.all_fields[fname]
      if x == nil then
        raise "Unkown field #{fname} @{@obj.schema_class}"
      else
        x.type
      end
    end
  end
  
  #Env that simulates the result of a lambda
  # can be used to store pointers
  class LambdaEnv
    include BaseEnv
    
    def initialize(label, parent = nil, &block)
      super()
      @label = label
      @block = block
      @parent = parent
    end
    
    def [](key)
      if @label == key
        @block.call
      else
        @parent && @parent[key]
      end
    end
    
    def []=(key, value)
      if @label == key
        raise "Trying to modify read-only variable #{key}"
      else
        @parent[key] = value
      end
    end
    
    def has_key?(key)
      @label == key || (@parent && @parent.has_key?(key))
    end

    def keys
      ([@label] + (@parent.nil? ? [] : @parent.keys)).uniq
    end
     
    def to_s
      "#{@block.to_s}-#{@parent}"
    end
  end
end

File: src/js/core/expr/code/env.js
'use strict'

//// Env ////

var cwd = process.cwd() + '/';
var Enso = require(cwd + "enso.js");

var Env;

function BaseEnv(parent) {
  return class extends parent {
    set_in_place(block, key) {
      var self = this;
      return self .set$(key, block(self.get$(key)));
    };

    set(block, key) {
      var self = this, res;
      res = self.clone();
      res.set_in_place(key, block);
      return res;
    };

    set_parent(env) {
      var self = this;
      self.parent$ = env;
      return self;
    };

    set_grandparent(env) {
      var self = this;
      if (self.parent$ == null || self.parent$ == Enso.EMap.new()) {
        return self.set_parent(env);
      } else {
        return self.parent$.set_grandparent(env);
      }
    };

    to_s() {
      var self = this, r;
      r = [];
      self.each(function(k, v) {
        return r.push(Enso.S(k, "=>", v));
      });
      return Enso.S("{ ", r.join(", "), " }");
    }; }};

class HashEnv extends Enso.mix(Enso.EnsoBaseClass, BaseEnv) {
  static new(...args) { return new HashEnv(...args) };

  constructor(hash = Enso.EMap.new(), parent = null) {
    super();
    var self = this;
    self.hash$ = hash;
    self.parent$ = parent;
  };

  get$(key) {
    var self = this, key;
    key = key.to_s();
    if (self.hash$.has_key_P(key)) {
      return self.hash$.get$(key);
    } else {
      return self.parent$ && self.parent$.get$(key);
    }
  };

  set$(key, value) {
    var self = this, key;
    key = key.to_s();
    if (self.hash$.has_key_P(key)) {
      return self.hash$ .set$(key, value);
    } else if (self.parent$ && self.parent$.has_key_P(key)) {
      return self.parent$ .set$(key, value);
    } else {
      return self.hash$ .set$(key, value);
    }
  };

  has_key_P(key) {
    var self = this, key;
    key = key.to_s();
    return self.hash$.has_key_P(key) || self.parent$ && self.parent$.has_key_P(key);
  };

  keys() {
    var self = this;
    return (self.hash$.keys() + (self.parent$ == null
      ? []
      : self.parent$.keys())).uniq();
  };

  to_s() {
    var self = this;
    return Enso.S(self.hash$.to_s(), "-", self.parent$);
  };
};

class ObjEnv extends Enso.mix(Enso.EnsoBaseClass, BaseEnv) {
  static new(...args) { return new ObjEnv(...args) };

  obj() { return this.obj$ };

  constructor(obj, parent = null) {
    super();
    var self = this;
    self.obj$ = obj;
    self.parent$ = parent;
  };

  get$(key) {
    var self = this;
    if (key == "self") {
      return self.obj$;
    } else if (self.obj$.schema_class().all_fields().any_P(function(f) {
      return f.name() == key;
    })) {
      return self.obj$.get$(key);
    } else {
      return self.parent$ && self.parent$.get$(key);
    }
  };

  set$(key, value) {
    var self = this;
    try {
      return self.obj$ .set$(key, value);
    } catch (DUMMY) {
      return self.parent$ && (self.parent$ .set$(key, value));
    }
  };

  has_key_P(key) {
    var self = this;
    return (key == "self" || self.obj$.schema_class().all_fields().get$(key)) || self.parent$ && self.parent$.has_key_P(key);
  };

  keys() {
    var self = this;
    return (self.obj$.schema_class().all_fields().keys() + (self.parent$ == null
      ? []
      : self.parent$.keys())).uniq();
  };

  to_s() {
    var self = this;
    return Enso.S(self.obj$.to_s(), "-", self.parent$);
  };

  type(fname) {
    var self = this, x;
    x = self.obj$.schema_class().all_fields().get$(fname);
    if (x == null) {
      return self.raise(Enso.S("Unkown field ", fname, " @{@obj.schema_class}"));
    } else {
      return x.type();
    }
  };
};

class LambdaEnv extends Enso.mix(Enso.EnsoBaseClass, BaseEnv) {
  static new(...args) { return new LambdaEnv(...args) };

  constructor(block, label, parent = null) {
    super();
    var self = this;
    self.label$ = label;
    self.block$ = block;
    self.parent$ = parent;
  };

  get$(key) {
    var self = this;
    if (self.label$ == key) {
      return self.block$();
    } else {
      return self.parent$ && self.parent$.get$(key);
    }
  };

  set$(key, value) {
    var self = this;
    if (self.label$ == key) {
      return self.raise(Enso.S("Trying to modify read-only variable ", key));
    } else {
      return self.parent$ .set$(key, value);
    }
  };

  has_key_P(key) {
    var self = this;
    return self.label$ == key || self.parent$ && self.parent$.has_key_P(key);
  };

  keys() {
    var self = this;
    return ([self.label$] + (self.parent$ == null
      ? []
      : self.parent$.keys())).uniq();
  };

  to_s() {
    var self = this;
    return Enso.S(self.block$.to_s(), "-", self.parent$);
  };
};

Env = {
  BaseEnv: BaseEnv,
  HashEnv: HashEnv,
  ObjEnv: ObjEnv,
  LambdaEnv: LambdaEnv,
};
module.exports = Env ;
